multi_imbalance/ensemble/ecoc.py

```diff
@@ -0,0 +1,285 @@
+        Returns
+        -------
+        self: object
+        """
+        self._labels = np.unique(y)
+        self._gen_code_matrix()
+        self._binary_classifiers = [self._get_classifier() for _ in range(self._code_matrix.shape[1])]
+        self._learn_binary_classifiers(X, y)
+        return self
+
+    def predict(self, X):
+        """
+        Parameters
+        ----------
+        X: two dimensional numpy array (number of samples x number of features) with float numbers
+        Returns
+        -------
+        y : numpy array, shape = [number of samples]
+            Predicted target values for X.
+        """
+        output_codes = np.zeros((X.shape[0], self._code_matrix.shape[1]))
+        for classifier_idx, classifier in enumerate(self._binary_classifiers):
+            output_codes[:, classifier_idx] = classifier.predict(X)
+
+        predicted = np.zeros(X.shape[0])
+        for row_idx, encoded_row in enumerate(output_codes):
+            predicted[row_idx] = self._get_closest_class(encoded_row)
+
+        return predicted
+
+    def _learn_binary_classifiers(self, X, y):
+        for classifier_idx, classifier in enumerate(self._binary_classifiers):
+            excluded_classes_indices = [idx for idx in range(len(y)) if
+                                        self._code_matrix[self._labels.tolist().index(y[idx])][classifier_idx] == 0]
+            X_filtered = np.delete(X, excluded_classes_indices, 0)
+            y_filtered = np.delete(y, excluded_classes_indices)
+            binary_labels = np.array([self._code_matrix[self._labels.tolist().index(clazz)][classifier_idx] for clazz in
+                                      y_filtered])
+            X_filtered, binary_labels = self._oversample(X_filtered, binary_labels)
+            classifier.fit(X_filtered, binary_labels)
+
+    def _gen_code_matrix(self):
+        if self.encoding == 'dense':
+            self._code_matrix = self._encode_dense(self._labels.shape[0])
+        elif self.encoding == 'sparse':
+            self._code_matrix = self._encode_sparse(self._labels.shape[0])
+        elif self.encoding == 'complete':
+            self._code_matrix = self._encode_complete(self._labels.shape[0])
+        elif self.encoding == 'OVO':
+            self._code_matrix = self._encode_ovo(self._labels.shape[0])
+        elif self.encoding == 'OVA':
+            self._code_matrix = self._encode_ova(self._labels.shape[0])
+        else:
+            raise ValueError("Unknown matrix generation encoding: %s, expected to be one of %s."
+                             % (self.encoding, ECOC._allowed_encodings))
+
+    def _encode_dense(self, number_of_classes, random_state=0, number_of_code_generations=10000):
+        try:
+            dirname = os.path.dirname(__file__)
+            matrix = np.load(dirname + f'/cached_matrices/dense_{number_of_classes}.npy')
+            return matrix
+        except:
+            pass
```

Comment:
We should catch some specific type instead of all exceptions and handle it in some way? At least, by printing error message ;)